cpp/src/arrow/compute/row/row_encoder_internal.cc

```diff
@@ -29,6 +29,51 @@ using internal::FirstTimeBitmapWriter;
 namespace compute {
 namespace internal {
 
+Result<std::shared_ptr<KeyEncoder>> MakeKeyEncoder(const TypeHolder& column_type, std::shared_ptr<ExtensionType>* extension_type, MemoryPool* pool) {
```

Comment:
This can also return `unique_ptr` here. I didn't see the purpose a shared_ptr being used